doc: add ability to generate PDF

dcpleung opened the pull request:
This adds new targets to generate build documentation through
LaTEX in PDF. Note that pdflatex complains about the tex file
generated by doxygen so it needs to be fixed with sed before
feeding in through pdflatex.

Relates to #6782.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

carlescufi commented on doc/CMakeLists.txt:
```diff
@@ -111,22 +116,83 @@ add_custom_target(
   # Merge the Doxygen and Sphinx logs into a single file
   COMMAND ${CMAKE_COMMAND} -P ${ZEPHYR_BASE}/cmake/util/fmerge.cmake ${DOC_LOG} ${DOXY_LOG} ${SPHINX_LOG}
   COMMAND ${PYTHON_EXECUTABLE} ${KI_SCRIPT} --config-dir ${CONFIG_DIR} --errors ${DOC_WARN} --warnings ${DOC_WARN} ${DOC_LOG}
+  COMMAND sed -z -i -f ${ZEPHYR_BASE}/doc/scripts/fix_tex/runaway_arg.sed ${SPHINX_OUTPUT_DIR_LATEX}/zephyr.tex
```
`sed` is UNIX-only. Can you use Python or CMake to achieve the same thing? We've just been through a painful exercise of making the doc build available on Windows

carlescufi commented:
@dcpleung thanks for this!
Could you please also update the macOS getting started instructions? you should be able to find the equivalent packages in brew. For Windows I can give it a go myself.

dcpleung commented:
The new commit has a Python script to do fix the tex file, and also has instructions for MacOS.

carlescufi commented on doc/CMakeLists.txt:
```diff
@@ -109,24 +114,86 @@ add_custom_target(
+# Hence the need of this to copy the PDF file over.
+add_custom_command(
+  OUTPUT ${SPHINX_OUTPUT_DIR_PDF}/zephyr.pdf
+  COMMAND mkdir -p -v ${SPHINX_OUTPUT_DIR_PDF}
```
Please use cmake's built-in functionality to create a folder and copy files to it:
https://cmake.org/cmake/help/v3.8/manual/cmake.1.html#command-line-tool-mode
`cmake -E ...`

This makes  sure this will work on all operating systems.

carlescufi commented on doc/CMakeLists.txt:
```diff
@@ -109,24 +114,86 @@ add_custom_target(
+add_custom_command(
+  OUTPUT ${SPHINX_OUTPUT_DIR_LATEX}/zephyr.pdf
+  DEPENDS latexdocs ${SPHINX_OUTPUT_DIR_LATEX}/zephyr.tex
+  COMMAND make LATEXOPTS="-halt-on-error" LATEXMKOPTS="-quiet"
```
what `Makefile` is this executing?

carlescufi replied:
Oh right, I see. This is generated by Sphinx itself. I think we can live with this for now, though long-term we need to find out if Sphinx is able to generate something else than Makefiles, since those won't work on Windows. I won't block the review because of this.

carlescufi commented:
@dcpleung thank you very much for helping make this cross-platform.
I  have added a couple more comments, since I think this is still UNIX-bound.

dcpleung committed:
doc: add ability to generate PDF

This adds new targets to generate build documentation through
LaTEX to PDF.

There are a few notes:

1. pdflatex complains about the tex file generated by doxygen
   so it needs to be fixed with a Python script before feeding
   in through pdflatex.

2. SVG files are not recognized by pdflatex so they are converted
   to known good format on the fly, only for producing PDF. This
   uses the libRSVG's rsvg-convert tool.

Relates to #6782.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dbkinder reviewed (changes requested):
With this PR, I'm getting an error with the normal ``htmldocs`` build target:

```
Extension error:
convert exited with error:
[stderr]
b"convert-im6.q16: unbalanced graphic context push-pop `graphic-context' @ error/draw.c/DrawImage/2305.\nconvert-im6.q16: non-conforming drawing primitive definition `pop' @ error/draw.c/DrawImage/3265.\n"
[stdout]
b''
ninja: build stopped: subcommand failed.
```
When I removed the ``sphinx.ext.imgconverter`` extension from ``conf.py``, the ``htmldocs`` target worked again without errors.  Not sure what's up, or which image conversion is the problem, but needs resolution before this can be merged since it breaks the website doc build.

It also looks like the CI system does not have the ``convert`` command (from  Imagemagick ) installed in its environment so it's throwing errors saying it can't find the convert command. See the shippable log.

dcpleung commented:
@carlescufi Thanks for the tips on cmake. I have converted those commands so now they should be portable.

@dbkinder I haven't seen that on my side. Which OS are you using? I can duplicate your setup and fix the issue.

dbkinder commented:
I'm on linux (ubuntu).  We'll also need to fix the CI environment (also linux).

dcpleung commented:
I have updated the patch to skip ImageMagick as it is highly dependent on the host system (for example, fonts installed vs. required to do svg->png).

dbkinder reviewed (changes requested):
That's cheating (with the .. only directives) but OK.  The ``htmldocs`` target works now.  But I get an error generating the PDF.  

I started with a ``ninja latexdocs`` (no errors) followed by a ``ninja pdfdocs``, but got this:

```
[5/6] Generating latex/zephyr.pdf
FAILED: latex/zephyr.pdf 
cd /home/dbkinder/zephyr/doc/_build/latex && /usr/bin/cmake -E env LATEXOPTS="-halt-on-error" /usr/bin/latexmk -quiet -pdf -dvi- -ps-
File::Glob::glob() will disappear in perl 5.30. Use File::Glob::bsd_glob() instead. at /usr/bin/latexmk line 7482.
Latexmk: Run number 1 of rule 'pdflatex'
This is pdfTeX, Version 3.14159265-2.6-1.40.18 (TeX Live 2017/Debian) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode

kpathsea: Running mktextfm ecrm1000
/usr/share/texlive/texmf-dist/web2c/mktexnam: Could not map source abbreviation  for ecrm1000.
/usr/share/texlive/texmf-dist/web2c/mktexnam: Need to update ?
mktextfm: Running mf-nowin -progname=mf \mode:=ljfour; mag:=1; nonstopmode; input ecrm1000
This is METAFONT, Version 2.7182818 (TeX Live 2017/Debian) (preloaded base=mf)

kpathsea: Running mktexmf ecrm1000

! I can't find file `ecrm1000'.
<*> ...ljfour; mag:=1; nonstopmode; input ecrm1000
                                                  
Please type another input file name
! Emergency stop.
<*> ...ljfour; mag:=1; nonstopmode; input ecrm1000
                                                  
Transcript written on mfput.log.
grep: ecrm1000.log: No such file or directory
mktextfm: `mf-nowin -progname=mf \mode:=ljfour; mag:=1; nonstopmode; input ecrm1000' failed to make ecrm1000.tfm.
kpathsea: Appending font creation commands to missfont.log.
Collected error summary (may duplicate other messages):
  pdflatex: Command for 'pdflatex' gave return code 256
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs of latex/pdflatex.
ninja: build stopped: subcommand failed.
```

dcpleung commented:
@dbkinder Did you install the fonts (package texlive-fonts-recommended)? The reference to ecrm1000 usually points to missing fonts.

I found a solution to avoid ImageMagick. Hopefully this one works better. :)

dbkinder commented:
Looking better... I did a fresh pull of your PR changes, verified the fonts package you mention was installed, along with the changed requirements.txt item.  The ``htmldocs`` target ran fine, as did the ``latexdocs`` target.  The ``pdfdocs`` target had some cryptic messages:

```
[5/6] Generating latex/zephyr.pdf
File::Glob::glob() will disappear in perl 5.30. Use File::Glob::bsd_glob() instead. at /usr/bin/latexmk line 7482.
Latexmk: Run number 1 of rule 'pdflatex'
This is pdfTeX, Version 3.14159265-2.6-1.40.18 (TeX Live 2017/Debian) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
Latexmk: Summary of warnings:
  Latex failed to resolve 1 reference(s)
Latexmk: Run number 1 of rule 'makeindex zephyr.idx'
Latexmk: Run number 2 of rule 'pdflatex'
This is pdfTeX, Version 3.14159265-2.6-1.40.18 (TeX Live 2017/Debian) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
Latexmk: Summary of warnings:
  Latex found 1 multiply defined reference(s)
Latexmk: Run number 3 of rule 'pdflatex'
This is pdfTeX, Version 3.14159265-2.6-1.40.18 (TeX Live 2017/Debian) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
Latexmk: Summary of warnings:
  Latex found 1 multiply defined reference(s)
[6/6] Generating pdf/zephyr.pdf
```
but did generate a PDF file, yea.  The "restricted \write18 enabled" message is described in https://tex.stackexchange.com/questions/76105/what-does-restricted-write18-enabled-mean-and-why-does-texlive-keep-reporting/76107, but the other messages seem cryptic.

Looking at the PDF file itself, all the content seems to be down one chapter level.  Instead of the chapters being the top-level TOC items, they're all generated within an unnecessary "Chapter 1".  This can be fixed by removing the "Sections" title before the top-level toc-tree in index.rst, but we can do that later, maybe under a ``.. only html`` tag.

dbkinder commented:
and the CI system is failing because of the missing extension (not installed)...

```
Could not import extension sphinxcontrib.rsvgconverter (exception: No module named 'sphinxcontrib.rsvgconverter')
```

dcpleung commented:
The latest commit fixes the write18 message and moves section headers up one level for non-html output. Also, the image converter is only enabled when needed since we don't need it for HTML.

dbkinder commented on doc/CMakeLists.txt:
```diff
@@ -109,24 +124,96 @@ add_custom_target(
   COMMAND ${CMAKE_COMMAND} -E env
   ZEPHYR_BUILD=${CMAKE_CURRENT_BINARY_DIR}
-  ${SPHINXBUILD} -w ${SPHINX_LOG} -N -t ${DOC_TAG} -b html ${ALLSPHINXOPTS} ${RST_OUT}/doc ${SPHINX_OUTPUT_DIR}
+  ${SPHINXBUILD} -w ${SPHINX_LOG} -N -t ${DOC_TAG} -b html -j auto ${ALLSPHINXOPTS} ${RST_OUT}/doc ${SPHINX_OUTPUT_DIR_HTML}
```
Can't use the ``-j auto`` option for parallel builds (yet).  That's the cause of the current CI failure about ``WARNING: Duplicate declaration, also defined in...``

See https://github.com/michaeljones/breathe/issues/141

carlescufi replied:
Wow, I missed that. Thanks @dbkinder

dbkinder commented on doc/CMakeLists.txt:
```diff
@@ -109,24 +124,96 @@ add_custom_target(
+  latex
+  COMMAND ${CMAKE_COMMAND} -E env
+  ZEPHYR_BUILD=${CMAKE_CURRENT_BINARY_DIR}
+  ${SPHINXBUILD} -w ${SPHINX_LOG} -N -t ${DOC_TAG} -b latex -t svgconvert -j auto ${ALLSPHINXOPTS} ${RST_OUT}/doc ${SPHINX_OUTPUT_DIR_LATEX}
```
remove ``-j auto`` option here too

dcpleung commented:
Removed the '-j auto' options.

dbkinder reviewed (approved):
OK by me now.  We'll likely need some tweaks to the PDF output in a followup PR. (I notice some links are going to the website instead of to the appropriate spot within the PDF because of how the links are written in the reST files, but it's close enough for now.)  Thanks.

dbkinder commented:
We should also (in another PR) update the doc building instructions to show how to build a PDF (/doc/README.rst)

marc-hb commented on doc/conf.py:
```diff
@@ -44,6 +44,10 @@
     'zephyr.application',
 ]
 
+# Only use SVG converter when it is really needed, e.g. LaTeX.
```
BTW I started using DOT_IMAGE_FORMAT = svg for Doxygen HTML output everywhere I can and I never looked back. Faster compile times, zoomable output and much, MUCH smaller disk usage. Doxygen hasn't changed the default yet because they don't want to break... IE9?

Just FYI.
